src/testdir/test_delete.vim

```diff
@@ -97,3 +97,48 @@ func Test_symlink_recursive_delete()
   call assert_equal(0, delete('Xdir4/Xfile'))
   call assert_equal(0, delete('Xdir4', 'd'))
 endfunc
+
+func Test_complicated_name_recursive_delete()
+  call mkdir('Xcomplicated')
+  call mkdir('Xcomplicated/[complicated-1 ]')
+  call mkdir('Xcomplicated/{complicated,2 }')
+  split Xcomplicated/Xfile
+  call setline(1, ['a', 'b'])
+  w
+  w Xcomplicated/\[complicated-1\ \]/Xfile
+  w Xcomplicated/\{complicated,2\ \}/Xfile
```

Comment:
Yes. But I tested only these two lines.
Actually, `w Xcomplicated/\[complicated-1\ ]/Xfile` causes `E16`, but the file was created as expected.
I don't know how to avoid `E16`. (Using `writefile()` is safer?)

I haven't tried rest of the patch.